radicle-cob/src/sign/signatures.rs

```diff
@@ -0,0 +1,291 @@
+// Copyright © 2019-2020 The Radicle Foundation <hello@radicle.foundation>
+//
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+use std::{
+    collections::BTreeMap,
+    convert::TryFrom,
+    iter::FromIterator,
+    ops::{Deref, DerefMut},
+};
+
+use git_trailers::{self as trailer, Token, Trailer};
+use serde::{
+    de::{value::StrDeserializer, IntoDeserializer},
+    Deserialize,
+};
+use trailer::OwnedTrailer;
+
+pub mod error;
+
+pub const TRAILER_TOKEN: &str = "X-Rad-Signature";
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Signature<Key, Sig> {
+    key: Key,
+    sig: Sig,
+}
+
+impl<K, S> From<(K, S)> for Signature<K, S> {
+    fn from((key, sig): (K, S)) -> Self {
+        Self { key, sig }
+    }
+}
+
+impl<K, S> TryFrom<&OwnedTrailer> for Signature<K, S>
+where
+    for<'de> K: Deserialize<'de>,
+    for<'de> S: Deserialize<'de>,
+{
+    type Error = error::Signature;
+
+    fn try_from(trailer: &OwnedTrailer) -> Result<Self, Self::Error> {
+        Self::try_from(Trailer::from(trailer))
+    }
+}
+
+impl<K, S> TryFrom<Trailer<'_>> for Signature<K, S>
+where
+    for<'de> K: Deserialize<'de>,
+    for<'de> S: Deserialize<'de>,
+{
+    type Error = error::Signature;
+
+    fn try_from(Trailer { values, .. }: Trailer) -> Result<Self, Self::Error> {
+        let mut iter = values.iter().flat_map(|val| val.split_whitespace());
+
+        let key = iter
+            .next()
+            .ok_or(error::Signature::Missing("public key"))
+            .and_then(|key| {
+                K::deserialize(
+                    key.deref().into_deserializer() as StrDeserializer<serde::de::value::Error>
+                )
```

Comment:
Are we using json inside commit messages? :grimacing:

Reply:
No, it's just the strings :) serde does not imply json ;)

Reply:
Ah, so deserializing strings into some type T via the Deserialize instance? Perhaps we can use FromStr/ToString here if I understand correctly.

Reply:
It doesn't matter now that we've agreed on `radicle-crypto` ;)

Reply:
Oh true.